pkg/types/types.go

```diff
@@ -44,6 +44,15 @@ type Registry interface {
 	CheckUpkeep(context.Context, bool, ...UpkeepKey) (UpkeepResults, error)
 }
 
+// Executer provides a runtime layer for parallelized execution of upkeeps,
+// with a cache for executed/pending upkeeps.
+// It acts as a mediator between observers and registry contract / mercury lookup.
+type Executer interface {
+	// Run executes the pipeline for the given upkeeps and collects the results.
+	// checkData is changed based on trigger.
+	Run(ctx context.Context, upkeepKeys []UpkeepKey, checkData [][]byte) (UpkeepResults, error)
```

Comment:
For the return, let's do []types.UpkeepResult instead of UpkeepResults unless we really need to extend the base array with some functionality.

Also, how should we handle errors here? If one upkeep check fails due to a 404 on mercury, do we run a retry on the upkeep? How soon do we fail for any specific upkeep?

Reply:
Some retry strategy could work, but we need to distinguish errors that are retry-able.
  In any case, currently the [error handling](https://github.com/smartcontractkit/chainlink/blob/develop/core/services/ocr2/plugins/ocr2keeper/evm/registry.go#L530) and mercury lookup are bound to the `EvmRegistry` so it might be better to handle things is that level.